Keywords unlocked part 1

bramtayl opened the pull request:
Part 1 of part 1 of #20402 for the recommendations here https://github.com/JuliaLang/julia/issues/25188#issuecomment-353028482 less qrfact[!]

bramtayl committed:
keyword for world in method_exists

bramtayl committed:
repeat keyword for Timer

bramtayl committed:
keywords for all, imported in names

bramtayl committed:
syntax keyword for code_native

bramtayl committed:
expand_generated keyword for code_lowered

bramtayl committed:
Revert "expand_generated keyword for code_lowered"

This reverts commit 377fe7056a1c6d1a3513533879655185ad855a19.

bramtayl committed:
base, pad keywords for digits

bramtayl committed:
base keyword for digits!

bramtayl committed:
overlap keyword for eachmatch

bramtayl committed:
overlap keyword for matchall

bramtayl committed:
head, tail keywords for chop

bramtayl committed:
base keyword for tryparse

bramtayl committed:
base keyword for parse

bramtayl committed:
start, stop keywords for reverse!

bramtayl committed:
start, stop keywords for reverse

bramtayl committed:
trim keyword for dropzeros

bramtayl committed:
trim keyword for dropzeros!

bramtayl committed:
mode keyword for mkdir

bramtayl committed:
mode keyword for mkpath

bramtayl committed:
mode keyword for chown

bramtayl committed:
eol keyword for countlines

bramtayl committed:
maxsize keyword for PipeBuffer

bramtayl committed:
own keyword for unsafe wrap

bramtayl commented:
The example for optional arguments that uses `parse` needs to be updated. I find the unixisms of mkpath/mkdir/chown particularly ugly. Any help tracking down bugs/omissions would be appreciated.

vtjnash commented on base/file.jl:
```diff
@@ -81,37 +81,40 @@ Temporarily changes the current working directory and applies function `f` befor
 cd(f::Function) = cd(f, homedir())
 
+checkmode(mode::Unsigned) = mode
+checkmode(mode::Signed) = throw(ArgumentError("mode must be an unsigned integer; try 0o$mode"))
```
this seems unnecessarily strict. maybe just drop the error? (convert will check InexectError anyways)

JeffBezanson replied:
It also shouldn't suggest `0o` this way because `mode` won't be printed in octal.

StefanKarpinski replied:
It probably makes sense to check that a mode only has three octal digits, but otherwise I think that any integer value between `0` and `511 = 0o777` should be acceptable, regardless of type.

vtjnash commented:
This is looking quite nice!

ararslan commented on base/pointer.jl:
```diff
@@ -68,7 +68,7 @@ unsafe_convert(::Type{Ptr{T}}, a::AbstractArray{T}) where {T} = error("conversio
 # unsafe pointer to array conversions
 """
-    unsafe_wrap(Array, pointer::Ptr{T}, dims, own=false)
+    unsafe_wrap(Array, pointer::Ptr{T}, dims; own = false)
```
Please don't change the whitespace around `=` for everything. Omitting it for keyword arguments is standard style in Base.

vtjnash replied:
I think it's fine

KristofferC replied:
I personally like it too, but adding such a change in this PR is quite unfortunate because it really invites bikeshedding.

ararslan replied:
Yes, if we could keep this to implementation only, that would be nice.

StefanKarpinski replied:
I think it's fine. It would have been preferable to do this without style changes and make style changes separately, but @bramtayl has been kind enough to do a large amount of tedious but important work here, so rather than asking him to do more, let's just focus on getting the meaningful content correct here and we can make our coding styles consistent to our hearts content after 1.0 is out.

JeffBezanson commented:
Seconded, this looks really good!

StefanKarpinski commented on base/file.jl:
```diff
@@ -616,12 +616,12 @@ function chmod(path::AbstractString, mode::Integer; recursive::Bool=false)
+    chown(path::AbstractString, owner::Integer; group::Integer = -1)
 
 Change the owner and/or group of `path` to `owner` and/or `group`. If the value entered for `owner` or `group`
 is `-1` the corresponding ID will not change. Only integer `owner`s and `group`s are currently supported.
```
I wonder if we should change this to use `nothing` as the sentinel value instead. We would have to translate `nothing` to -1 for libuv, but using -1 just because libuv does seems to be exposing an implementation detail. So the behavior here would be to error on negative values, replace `nothing` with `-1`, and pass positive integers through unchanged. Not a big deal, just a thought now that we use `nothing` commonly as a sentinel.

bramtayl replied:
That makes more sense to me too.

StefanKarpinski commented on base/loading.jl:
```diff
@@ -88,7 +88,7 @@ macro return_if_file(path)
 
 function find_package(name::String)
-    endswith(name, ".jl") && (name = chop(name, 0, 3))
+    endswith(name, ".jl") && (name = chop(name, head = 0, tail = 3))
```
Here you can drop the `head = 0` since that's the default. It was only needed for the positional form because the `head` argument is first.

fredrikekre replied:
^ bump

StefanKarpinski commented on base/event.jl:
```diff
@@ -341,7 +341,7 @@ end
 
 """
-    Timer(delay, repeat=0)
+    Timer(delay; repeat = 0)
```
I think the `repeat` argument to `Timer` is kind of confusingly named – it seems like it gives a repeat count, but instead gives the interval of repetition. Perhaps it should be called `interval` instead with the same behavior? I.e. if given the timer repeats with the given interval, if not given the timer does not repeat. We might in the future want to add a `repeat` keyword argument which controls how many _times_ a timer repeats. Of course, we might want to call such an argument `count` instead, but you see my point.

bramtayl replied:
I was confused by `repeat` at first too. I'll change it to `interval`.

StefanKarpinski commented on base/deprecated.jl:
```diff
@@ -3806,6 +3806,32 @@ end
+@deprecate chop(s, head) chop(s, head = head)
+@deprecate chop(s, head, tail) chop(s, head = head, tail = tail)
+@deprecate tryparse(T, s, base) tryparse(T, s, base = base)
+@deprecate parse(T, s, base) parse(T, s, base = base)
```
Should these have `T::Type{<:Integer}`or something like that?

bramtayl replied:
Sure, that sounds right

StefanKarpinski reviewed (commented):
This is a really pleasant change! Thanks so much for taking the time and effort to go through this so thoroughly. I've left a few comments which should be addressed (the `Timer` / `repeat` one is the most significant) but aside from those issues, this looks great to me.

nalimilan commented on base/deprecated.jl:
```diff
@@ -3806,6 +3806,32 @@ end
+@deprecate reverse!(v, start) reverse!(v, start = start)
+@deprecate reverse!(v, start, stop) reverse!(v, start = start, stop = stop)
+@deprecate reverse(v, start) reverse(v, start = start)
+@deprecate reverse(v, start, stop) reverse(v, start = start, stop = stop)
```
I wonder whether `reverse` should be kept as is. There's a pattern of passing start and stop indices as keyword arguments, e.g. with `SubString` and `copy!`. I prefer using keyword arguments for arguments which are of a "different nature" than other arguments (typically, options vs. data).

bramtayl replied:
I don't understand...

nalimilan replied:
What don't you understand exactly? :-)

mbauman replied:
@nalimilan did you typo "There's a pattern of passing start and stop indices as keyword argument"?  I think you meant ~keyword~ positional argument.

nalimilan replied:
Ah, thanks, that explains it!

StefanKarpinski replied:
I would be fine with leaving `reverse` and `reverse!` as they are but I don't hate this either.

bramtayl replied:
I'm not super passionate either but its worth noting that if they are both optional arguments than you wouldn't be able to do something like `reverse(v, stop = stop)`

nalimilan replied:
Yeah, you would have to pass `start(v)` before the stop index. But this remark also applies to functions with similar arguments, so I think we should keep them consistent (either way).

bramtayl replied:
Do you have a list of such functions? It might be just worth changing all of them (if they are exported).

StefanKarpinski replied:
How about dropping that part for now so that we can get this merged now?

quinnj commented:
The only bummer of this is we still have https://github.com/JuliaLang/julia/issues/5230, so this introduces a whole swath of new methods that you basically can't see `@code_*` for.

StefanKarpinski commented:
Fortunately, that can be fixed at any point and no one will complain.

StefanKarpinski commented:
@bramtayl, could you remove the `reverse` and `reverse!` changes here and rebase so that we can get this merged?

bramtayl committed:
Revert "start, stop keywords for reverse!"

This reverts commit baba473c55b2c300bb38e1acf6eae7e428a347b7.

bramtayl committed:
Revert "start, stop keywords for reverse"

This reverts commit f9c3b308ce4a2f4ea25c4b8a2e07a6fd2c8e65fe.

bramtayl committed:
updates to checkmode

bramtayl committed:
tighted (try)parse deprecations

bramtayl committed:
repeat -> interval for Timer

bramtayl committed:
missing, nothing as sentinels in chown

bramtayl commented:
Sure thing. I've pushed some commits that addressed the comments.

For `chown` it seemed to me from the tests that there was a distinction between -1 (don't change) and -2 (remove). I've updated these to use sentinel values of `missing` and `nothing`. I'm guessing this change would be better in a separate PR?

I still need
1) To replace the example for optional arguments that uses `parse`. Any suggestions?
2) Rebase, test, bugfix

Sorry for moving slowly on this. I clearly do not have the attention span for fiddly PRs.

nalimilan commented:
> For chown it seemed to me from the tests that there was a distinction between -1 (don't change) and -2 (remove). I've updated these to use sentinel values of missing and nothing. I'm guessing this change would be better in a separate PR?

Yes, I don't think we should use `missing` for this. It's purpose is to represent a missing value. Better keep only uncontroversial changes in the PR so that it can be merged.

StefanKarpinski reviewed (commented):
A few small issues, but overall looking great. If you want, I can make these changes to the branch or let you make them, but I'd like to get this merged ASAP. Let me know whether I should do it or you will.

bramtayl committed:
fix default for chown

bramtayl committed:
fix timer bug

bramtayl commented:
Well, I think I just addressed the bugs you mentioned, but also feel free make any changes you see fit.

bramtayl committed:
Merge branch 'master' of github.com:JuliaLang/julia into keywords_unlocked

StefanKarpinski commented:
I'm mulling over the `chown` signature. I like the use of `missing` and `nothing` value, but it just feels weird to have the `owner` permission value be positional but the `group` value be keyword. The two options without that inconsistency are:
1. make both `owner` and `group` permissions positional with `group` being optional, defaulting to `missing` which means "don't change". Examples:
```jl
chown("file.txt", 123) # change owner to user 123, leave group alone
chown("file.txt", 123, 9) # change owner to user 123, change group to 9
chown("file.txt", missing, 9) # change group to 9, leave owner alone
```
2. make both `user` and `group` permissions keyword arguments defaulting to `missing`. Examples:
```jl
chown("file.txt", owner=123) # change owner to user 123, leave group alone
chown("file.txt", owner=123, group=9) # change owner to user 123, change group to 9
chown("file.txt", group=9) # change group to 9, leave owner alone
```

At first I liked the use of `nothing` and `missing` here, but I've changed my mind. As @nalimilan points out, it's an abuse of both `missing` – which means "there is no value" – and of `missing` which means "there is a value but we do not have it", and is intended for representing missing data.

bramtayl commented:
I like option 2, I think.

nalimilan commented on base/file.jl:
```diff
@@ -616,13 +620,36 @@ function chmod(path::AbstractString, mode::Integer; recursive::Bool=false)
+entered for `owner` or `group` is `missing` the corresponding ID will not
+change. If the value entered for `owner` or `group` is `nothing` the
+corresponding ID will be `nobody` or `nogroup` respectively. Only non-negative
+integer `owner`s and `group`s are currently supported.
```
You think, or you're sure? :-)

I'm sure I don't like it. We never use `missing` as an argument value to change the behavior of a function. I'd rather not add this here and now as we don't really have the time to discuss a more general design about where this kind of pattern would or wouldn't be OK.

StefanKarpinski commented:
Option 3: add a separate `chgrp` function which takes a file and a single, required positional argument that changes the group of its argument. That said, I think I favor option 2 as well.

StefanKarpinski commented:
Yes, I think we should probably do that, although I suspect that using symbols for special meanings will be the way to go, e.g. `chown("file.txt", owner=:nobody, group=:nogroup)`. Or, if we want to support looking up users and groups by name and the system understands `"nobody"` and `"nogroup"` as names, then one could just use those strings for this.

bramtayl commented:
Maybe I was confused, but from the tests it looks like we support both -1 and -2, that is, do nothing and nobody/nogroup.

StefanKarpinski commented:
Previously we supported it just by passing values through to libuv – the issue is just that magical negative constants are a bit weird, which we're trying to fix here.

bramtayl commented:
Rihgt. Well I'm not super passionate about this so I'd go with whatever. Worth noting that `chown` is never used in `Base` so it might be nice as part of a stdlib?

StefanKarpinski commented:
> Worth noting that `chown` is never used in Base so it might be nice as part of a stdlib?

Ah, that is an interesting point.

JeffBezanson commented:
Let's leave alone any functions where there are significant questions, so we can get the rest in.

bramtayl committed:
Revert "fix default for chown"

This reverts commit 095bbd4f52bda85e3549321b69b9d3e2c73b9695.

bramtayl committed:
Revert "missing, nothing as sentinels in chown"

This reverts commit 2c01c4d333ea935f6923a2190bec60dc51a07d12.

bramtayl committed:
Revert "mode keyword for chown"

This reverts commit 8903628453ee7ba249ffe7b6a4cbcc67854041f9.

bramtayl committed:
Merge branch 'master' into keywords_unlocked

bramtayl commented:
Ok I removed the `chown` commits and rebased.

fredrikekre commented on base/deprecated.jl:
```diff
@@ -2801,6 +2801,28 @@ function Filesystem.tempname(uunique::UInt32)
+@deprecate tryparse(T::Type{<:Integer}, s, base) tryparse(T, s, base = base)
+@deprecate parse(T::Type{<:Integer}, s, base) parse(T, s, base = base)
+@deprecate dropzeros(x, trim) dropzeros(x, trim = trim)
+@deprecate dropzeros!(x, trim) dropzeros!(x, trim = trim)
```
`dropzeros` and `dropzeros!` should go into `stdlib/SparseArrays/src/deprecate.jl`

fredrikekre commented on base/deprecated.jl:
```diff
@@ -586,14 +586,14 @@ end
 
 # PR #22088
 function hex2num(s::AbstractString)
-    depwarn("`hex2num(s)` is deprecated. Use `reinterpret(Float64, parse(UInt64, s, 16))` instead.", :hex2num)
```
The backtics should stay

fredrikekre commented on base/event.jl:
```diff
@@ -341,12 +341,12 @@ end
 
 Waiting tasks are woken after an intial delay of `delay` seconds, and then repeating with the given
-`repeat` interval in seconds. If `repeat` is equal to `0`, the timer is only triggered once. When
+`interval` in seconds. If `repeat` is equal to `0`, the timer is only triggered once. When
```
`` `repeat` `` -> `` `interval` ``

fredrikekre commented on base/file.jl:
```diff
@@ -81,37 +81,44 @@ Temporarily changes the current working directory and applies function `f` befor
 
+function checkmode(mode::Int) =
+    if !(0 <= mode <= 511)
+        ArgumentError("Mode must be between 0 and 511 = 0o777")
```
`throw(...)` ?

fredrikekre commented on base/parse.jl:
```diff
@@ -195,29 +195,23 @@ function tryparse_internal(::Type{Bool}, sbuff::Union{String,SubString{String}},
 
 """
-    tryparse(type, str, [base])
+    tryparse(type, str, base = base)
```
`base = 0`?

bramtayl committed:
backticks

bramtayl committed:
`repeat` -> `interval`

bramtayl committed:
throw

bramtayl committed:
remove head = 0

bramtayl committed:
base = 0

bramtayl committed:
move dropzeros to stdlib

bramtayl commented:
Thanks everyone for the detailed comments! I fixed some more things.

bramtayl committed:
Merge branch 'master' into keywords_unlocked

bramtayl committed:
Merge branch 'master' into keywords_unlocked

JeffBezanson commented on base/strings/util.jl:
```diff
@@ -85,15 +85,15 @@ julia> a = "March"
 """
 chop(s::AbstractString) = SubString(s, start(s), prevind(s, endof(s)))
-chop(s::AbstractString, head::Integer, tail::Integer) =
+chop(s::AbstractString; head::Integer, tail::Integer) =
```
Forgot the default values here.

JeffBezanson commented:
Rebasing this now should give faster build times.

bramtayl committed:
defaults for chop

bramtayl committed:
Merge branch 'master' into keywords_unlocked

bramtayl committed:
Merge branch 'keywords_unlocked' of https://github.com/bramtayl/julia into keywords_unlocked

bramtayl committed:
removed extraneous =

fredrikekre commented on base/strings/util.jl:
```diff
@@ -85,15 +85,15 @@ julia> a = "March"
 """
 chop(s::AbstractString) = SubString(s, start(s), prevind(s, endof(s)))
-chop(s::AbstractString, head::Integer, tail::Integer) =
+chop(s::AbstractString; head::Integer = 0, tail::Integer = 1) =
```
This will overwrite the method on the line above, right?

bramtayl replied:
Oh shoot I messed this up. If I understand correctly `start(s)` and `prevind(s, endof(s))` are supposed to be optimizations for the default cause. If I'm understanding correctly constant propagation is not working right now for keyword arguments. So maybe I should comment out the first method and add back in the optimization if that ever happens?

bramtayl committed:
remove overwritten chop

bramtayl committed:
sp

bramtayl committed:
import mkpath

bramtayl committed:
missing end

JeffBezanson commented on base/deprecated.jl:
```diff
@@ -2819,6 +2819,27 @@ export readandwrite
+@deprecate parse(T::Type{<:Integer}, s, base) parse(T, s, base = base)
+@deprecate mkdir(path, mode) mkdir(path, mode = mode)
+import Filesystem.mkpath
+@deprecate mkpath(path, mode) mkpath(path, mode = mode)
```
Can wrap this line in `@eval Filesystem begin ... end` instead.

bramtayl committed:
eval Filesystem

bramtayl committed:
widen checkmode

bramtayl committed:
eval filesystem for mkdir

bramtayl commented:
I'm having a heck of a time debugging this while rebasing it at the same time

JeffBezanson commented:
Rebased and fixed some things: #25647
